File: Actions/ConfigFiles.hs
{- dfsbuild: CD image builder
Copyright (c) 2006 John Goerzen
Please see COPYRIGHT for more details
-}

module Actions.ConfigFiles where
import Utils
import qualified Actions.Mirror
import System.Posix.Directory
import System.Posix.Files
import Data.String.Utils
import System.Cmd.Utils
import System.Path
import System.Path.Glob
import System.FilePath
import Control.Monad
import Data.ConfigFile
import System.IO.HVFS
import System.Time
import Text.Printf

getDate =
    getClockTime >>= toCalendarTime >>= (return . calendarTimeToString)

buildinfo env =
    return $ "This is a Debian From Scratch (DFS) live CD made by dfsbuild" ++
                "\n\nName: " ++ eget env "name" ++
                "\nVersion: " ++ eget env "version" ++
                "\nBuilder: " ++ eget env "builder" ++
                "\nPreparation Date: " ++ datestr env ++ "\n"

writeBuildInfo env = 
    buildinfo env >>= 
                  (writeFile ((targetdir env) ++ "/opt/dfsruntime/buildinfo"))

getidstring :: DFSEnv -> String
getidstring env = 
    printf "DFS: %s %s (%s)" (eget env "name") (eget env "version")
           (datestr env)
    
                     
writeCfgFiles env =
 do bi <- buildinfo env
    writeit appendFile "/etc/issue" ("\n" ++ bi ++ "\n")
    case options (cp env) "appendfiles" of
      Left _ -> return ()
      Right files ->
           mapM_ (\fn -> writeit appendFile fn (esget env "appendfiles" fn))
                 files
    case options (cp env) "createfiles" of
      Left _ -> return ()
      Right files ->
          mapM_ (\fn -> writeit writeFile fn (esget env "createfiles" fn))
                files
    case get (cp env) (defaultArch env) "deletefiles" of
      Left _ -> return ()
      Right files ->
          do delfiles <- mapM glob (map ((targetdir env) ++) (splitWs files))
             mapM_ deleteit $ concat delfiles
    case options (cp env) "symlinks" of
      Left _ -> return ()
      Right files ->
          mapM_ (\from -> do dm $ "Symlinking " ++ from
                             createSymbolicLink (esget env "symlinks" from)
                                                ((targetdir env) ++ from)
                ) files
    case get (cp env) (defaultArch env) "makedirs" of
      Left _ -> return ()
      Right files ->
          mapM_ (\fn -> createDirectory ((targetdir env) ++ fn) 0o755)
                (splitWs files)

 where writeit func fn info =
           func ((targetdir env) ++ fn) (info ++ "\n")

fixRc env =
    do recursiveRemove SystemFS ((targetdir env) ++ "/etc/rc2.d")
       safeSystem "cp" ["-r", targetdir env ++ "/etc/rc1.d",
                        targetdir env ++ "/etc/rc2.d"]
       cpfiles <- glob $ targetdir env ++ "/etc/rc3.d/*logd*"
       safeSystem "cp" $ ["-r"] ++  cpfiles ++ [targetdir env ++ "/etc/rc2.d/"]
       rmfiles <- glob $ targetdir env ++ "/etc/rc2.d/S*single"
       mapM_ deleteit rmfiles


kernelimgconf =
    "do_symlinks = no\n\
\do_bootloader = no\n\
\do_bootfloppy = no\n\
\do_initrd = yes\n\
\warn_initrd = yes\n"

File: old-ocaml/configfiles.ml
(* arch-tag: write config files
*)
open Dfsutils;;
open Shellutil;;
open Unixutil;;
open Strutil;;
open Archsupport;;

let datestr = strip (run_getstring "date" []);;

let getbuildinfo cp = 
  "Name: " ^ (get cp "name") ^ 
  "\nVersion: " ^ (get cp "version") ^
  "\nBuilder: " ^ (get cp "builder") ^
  "\nPreparation Date: " ^ datestr ^ "\n";;

let writebuildinfo cp target =
  let fd = open_out (target ^ "/opt/dfsruntime/buildinfo") in
  output_string fd (getbuildinfo cp);
  Pervasives.close_out fd;;

let getidstring cp =
  Printf.sprintf "DFS image: %s %s (%s)" (get cp "name") (get cp "version") datestr;;

let writecfgfiles cp basedir =
  let w fn s =
    let outfd = open_out (basedir ^ fn) in
    output_string outfd s;
    output_string outfd "\n";
    close_out outfd
  in
  let a fn s =
    let outfd = open_out_gen [Open_append; Open_wronly; Open_creat] 0o644
    (basedir ^ fn) in
    output_string outfd s;
    output_string outfd "\n";
    close_out outfd
  in

  a "/etc/issue" ("\n" ^ (getbuildinfo cp) ^ "\n");
  
  if cp#has_section "appendfiles" then begin
    List.iter (fun fn -> a fn (cp#get "appendfiles" fn))
     (cp#options "appendfiles");
  end;
  if cp#has_section "createfiles" then begin
    List.iter (fun fn -> w fn (cp#get "createfiles" fn))
      (cp#options "createfiles");
  end;
  if cp#has_section "symlinks" then begin
    List.iter (fun from -> p("Symlinking " ^ from); Unix.symlink (cp#get "symlinks" from) (basedir ^
    from))
      (cp#options "symlinks");
  end;
      List.iter (fun x -> p("Deleting " ^ x); rm ~force:true (basedir ^ x))
        (glob (split_ws (get ~default:"" cp "deletefiles")));
  p("Looking for dirs to make...");
  p (get cp "makedirs");
      List.iter (fun x -> p("Making dir " ^ x); Unix.mkdir (basedir ^ x) 0o755)
        (split_ws (get ~default:"" cp "makedirs"));
;;

let fixrc target =
  rm ~recursive:true (target ^ "/etc/rc2.d");
  run "cp" ["-r"; (target ^ "/etc/rc1.d"); (target ^ "/etc/rc2.d")];
  run "cp" ("-r" :: glob [target ^ "/etc/rc3.d/*logd*"] @
    [target ^ "/etc/rc2.d/"]);
  List.iter (fun x -> rm x) (glob [target ^ "/etc/rc2.d/S*single"]);
;;

